onnxruntime/test/global_thread_pools/test_inference.cc

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <core/common/make_unique.h>
+#include "core/session/onnxruntime_cxx_api.h"
+#include "core/graph/constants.h"
+#include "providers.h"
+#include <memory>
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <atomic>
+#include <gtest/gtest.h>
+#include "test_allocator.h"
+#include "../shared_lib/onnx_protobuf.h"
+#include "../shared_lib/test_fixture.h"
+#include <stdlib.h>
+
+struct Input {
+  const char* name = nullptr;
+  std::vector<int64_t> dims;
+  std::vector<float> values;
+};
+
+extern std::unique_ptr<Ort::Env> ort_env;
+static constexpr PATH_TYPE MODEL_URI = TSTR("testdata/squeezenet/model.onnx");
+class CApiTestGlobalThreadPoolsWithProvider : public testing::Test, public ::testing::WithParamInterface<int> {
+};
+
+template <typename OutT>
+static void RunSession(OrtAllocator* allocator, Ort::Session& session_object,
+                       const std::vector<Input>& inputs,
+                       const char* output_name,
+                       const std::vector<int64_t>& dims_y,
+                       const std::vector<OutT>& values_y,
+                       Ort::Value* output_tensor) {
+  std::vector<Ort::Value> ort_inputs;
+  std::vector<const char*> input_names;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    input_names.emplace_back(inputs[i].name);
+    ort_inputs.emplace_back(Ort::Value::CreateTensor<float>(allocator->Info(allocator), const_cast<float*>(inputs[i].values.data()), inputs[i].values.size(), inputs[i].dims.data(), inputs[i].dims.size()));
```

Comment:
>const_cast<float*>(inputs[i].values.data()) [](start = 88, length = 43)

Why are we doing a const_cast here?

Reply:
Copied from onnxruntime/test/shared_lib/test_inference.cc. CreateTensor expects a non-const ptr. I can make the inputs vector param non-const to avoid this, although it's weird since we're not really modifying the inputs vector.

Reply:
Technically if you put it in a Tensor someone can call MutableData and modify it. So it's actually a good hint that the inputs could potentially be modified. You could also copy the inputs into a local vector if that seems cleaner.

---
In reply to: [393469401](https://github.com/microsoft/onnxruntime/pull/3177#discussion_r393469401) [](ancestors = 393469401)